src/sqlfluff/core/rules/surrogates/segments.py

```diff
@@ -0,0 +1,93 @@
+"""Surrogate class for working with Segment collections."""
+from typing import Callable, List, Optional, Sequence, Type, TypeVar
+
+from sqlfluff.core.parser import BaseSegment
+from sqlfluff.core.rules.base import LintFix
+from sqlfluff.core.templaters.base import TemplatedFile
+from sqlfluff.core.rules.surrogates.raw_file_slice import RawFileSlices
+
+Predicate = TypeVar("Predicate", str, Type, Callable[[BaseSegment], bool])
+
+
+class Segments:
+    """Encapsulates a sequence of one or more BaseSegments.
+
+    The segments may or may not be contiguous in a parse tree.
+    Provides useful operations on a sequence of segments to simplify rule creation.
+    """
+
+    def __init__(self, templated_file: Optional[TemplatedFile], *segments: BaseSegment):
+        self.templated_file = templated_file
+        self.segments = segments
+
+    def all(self, *predicates: Predicate) -> bool:
+        """Do all the segments match?"""
+        cp = _CompositePredicate(*predicates)
+        return all(cp(s) for s in self.segments)
+
+    def any(self, *predicates: Predicate) -> bool:  # pragma: no cover
+        """Do any of the segments match?"""
+        cp = _CompositePredicate(*predicates)
+        return any(cp(s) for s in self.segments)
+
+    @property
+    def raw_slices(self) -> RawFileSlices:
+        """Raw slices of the segments."""
+        if not self.templated_file:  # pragma: no cover
+            raise ValueError(
+                'Segments.raw_slices: "templated_file" property is required.'
+            )
+        raw_slices = set()
+        for s in self.segments:
+            source_slice = s.pos_marker.source_slice
+            raw_slices.update(
+                self.templated_file.raw_slices_spanning_source_slice(source_slice)
+            )
+        return RawFileSlices(self.templated_file, *raw_slices)
+
+    # def with_children(self) -> "Segments":
+    #     """Returns an object that includes first-level children."""
+    #     raise NotImplementedError
+
+    # def with_descendants(self) -> "Segments":
+    #     """Returns an object that includes all descendants."""
+    #     raise NotImplementedError
+
+    # def select(
+    #     self,
+    #     start_seg: Optional[BaseSegment] = None,
+    #     stop_seg: Optional[BaseSegment] = None,
+    #     select_if=Optional[Sequence[Predicate]],
+    #     loop_while=Optional[Sequence[Predicate]],
+    # ) -> "Segments":
+    #     """Retrieve range/subset."""
+    #     raise NotImplementedError
```

Comment:
I assume this acts like a filter so I can say something like "return me the whitespace segments before next code segment" using something like:
```python
raw_stack = context.surrogates.raw_stack
leading_whitespace = raw_stack.select(select_if=[attr("is_whitespace")], loop_while=[attr("is_code")])
```

Reply:
One thing the above example got me thinking about is that it's hard to negate these predicates.

So if I want to select NOT whitespace I can't do
`raw_stack.select(select_if=[not attr("is_whitespace")])`
because that's not a callable.
And we don't have a `is_not_whitespace` attribute.

So I would have to make my own function to move the not within the function:
`raw_stack.select(select_if=[lambda x: not x.is_whitespace])`
which would work, but feels less natural than a list comprehension

Reply:
It should be easy to add a `not()` function to wrap and negate another predicate. I'll start a checklist in the PR description with a note to add this. 👍